### ISSUE TITLE
Nodejs - acting as client to other services (windows authentication)

### ISSUE THREAD

**Issue by MattWorkWeb** [nodejs/help#1089]
* **Node.js Version**:   6.11.3
* **OS**: Windows Server 12
* **Scope (install, code, runtime, meta, other?)**:
* **Module (and version) (if relevant)**:

 I’m working a problem this week and looking for nodejs experience/insight.  Problem:

1)	Our nodejs server running on prem on windows server 2016 has to make some rest calls to another on prem web service.
2)	The other web service has windows authentication enabled through IIS.
3)	We enabled windows authentication for our nodejs server but it still cannot communicate with the support service as it’s not passing through credentials (i.e., token info).  Any REST request I make to the other on premise service is 401/unauthorized.

So we see if our server was not nodejs but asp, asp has an option on WebClient to UseDefaultCredentials which enables pass through of windows authentication info from client to server service; we do not see options for this in nodejs web requests.  Any knowledge or guidance here?  We see some statements about rolling your own NTLM support and some packages (e.g., passport, sspi) but they still seem to require reauthentication (i.e., instead of simply passing through credentials known similar to asp WebClient UseDefaultCredentials).

Any thoughts?  THanks.

-->

**Comment by bnoordhuis** [comment]
> We see some statements about rolling your own NTLM support and some packages (e.g., passport, sspi) but they still seem to require reauthentication

That doesn't sound right.  Most authentication modules will let you use an existing token if you have one.  That's not radically different from `WebClient.DefaultCredentials`.

**Comment by MattWorkWeb** [comment]
How do the authentication modules let me use an existing token; I've tried (i.e., within node server) simply taking the token I receive in the request, and placing it in the auth header of my outgoing request (i.e., from node server to another service) but I'm still seeing 401/unauthorize.  The only recourse I see is to reauthorize using username and password?  I'll take any guidance.

**Comment by bnoordhuis** [comment]
What module and auth method are you wanting to use?

When you say you get a 401, do you mean for the first request?  You're supposed to use the `WWW-Authenticate` header from the 401 response to compute the `Authorization` header for the next request, and fetch a service ticket from the AS in between.

It's possible `WebClient` does that automatically.  Most node.js libraries probably don't.

**Comment by gireeshpunathil** [comment]
ping @MattWorkWeb

**Comment by sagitsofan** [comment]
Still needing a fix for that issue..

**Comment by gireeshpunathil** [comment]
@sagitsofan - can you try what @bnoordhuis has suggested in 
https://github.com/nodejs/help/issues/1089#issuecomment-362401219 ?

**Comment by sagitsofan** [comment]
I don't want to manually calculate and taking care of the "Authorization" header in every request.
I need a solution like just "UseDefaultCredentials " in .NET, but in my node server.
Just pass the identity of the current user.
Don't you know a good package that handles this logic?

**Comment by gireeshpunathil** [comment]
makes sense, but have no idea whether this is available in an `npm` module.

**Comment by sagitsofan** [comment]
:(

**Comment by gireeshpunathil** [comment]
can you please ask this question to https://github.com/expressjs/express ? It is not that they should know the answer, but it is really is a middleware question, not a runtime one. So they may be able to provide a better hint I guess.

**Comment by sagitsofan** [comment]
I will try, thank you!
